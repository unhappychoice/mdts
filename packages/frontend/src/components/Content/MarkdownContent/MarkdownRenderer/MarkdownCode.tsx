import React from 'react';
import { useSelector } from 'react-redux';
import { PrismAsyncLight } from 'react-syntax-highlighter';
import { SYNTAX_HIGHLIGHTER_THEMES } from '../../../../constants';
import { useSyntaxHighlighterTheme } from '../../../../hooks/useSyntaxHighlighterTheme';
import { RootState } from '../../../../store/store';
import MermaidRenderer from './MermaidRenderer';

interface MarkdownCodeProps {
  inline?: boolean;
  className?: string;
  children?: React.ReactNode;
}

const MarkdownCode: React.FC<MarkdownCodeProps> = ({ inline, className, children, ...props }) => {
  const { fontFamilyMonospace, syntaxHighlighterTheme } = useSelector((state: RootState) => state.config);
  const syntaxHighlighterStyle = useSyntaxHighlighterTheme(syntaxHighlighterTheme);

  const match = /language-(\w+)/.exec(className || '');
  if (match && match[1] === 'mermaid') {
    return <MermaidRenderer chart={String(children).replace(/\n$/, '')} />;
  }

  return !inline && match ? (
    <PrismAsyncLight
      style={syntaxHighlighterStyle}
      className={'syntax-highlighter'}
      customStyle={{ margin: 0 }}
      showLineNumbers={true}
      language={match[1]}
      codeTagProps={{ style: { fontFamily: fontFamilyMonospace } }}
      PreTag="div"
      {...props}
    >
      {String(children).replace(/\n$/, '')}
    </PrismAsyncLight>
  ) : (
    <code
      {...props}
      className={className}
      style={{
        color: SYNTAX_HIGHLIGHTER_THEMES.find(theme => theme.value === syntaxHighlighterTheme)?.type === 'light'
          ? 'rgba(0, 0, 0, .87)'
          : '#fff',
        fontFamily: fontFamilyMonospace,
        backgroundColor: SYNTAX_HIGHLIGHTER_THEMES.find(theme => theme.value === syntaxHighlighterTheme)?.color,
        borderRadius: '4px',
      }}
    >
      {children}
    </code>
  );
};

export default MarkdownCode;
