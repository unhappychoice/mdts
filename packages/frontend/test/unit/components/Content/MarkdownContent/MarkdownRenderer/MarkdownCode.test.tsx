import { act, render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import MarkdownCode from '../../../../../../src/components/Content/MarkdownContent/MarkdownRenderer/MarkdownCode';

const mockStore = configureStore([]);

describe('MarkdownCode', () => {
  let store: unknown;

  beforeEach(() => {
    store = mockStore({
      config: {
        fontFamilyMonospace: 'monospace',
        syntaxHighlighterTheme: 'github',
      },
    });
  });

  it('should render inline code correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MarkdownCode inline>{'console.log("hello");'}</MarkdownCode>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render code block correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MarkdownCode className="language-js">{'console.log("hello");'}</MarkdownCode>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render mermaid chart correctly', async () => {
    let asFragment: unknown;
    await act(async () => {
      ({ asFragment } = render(
        <Provider store={store}>
          <MarkdownCode className="language-mermaid">{'graph TD; A-->B'}</MarkdownCode>
        </Provider>
      ));
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
