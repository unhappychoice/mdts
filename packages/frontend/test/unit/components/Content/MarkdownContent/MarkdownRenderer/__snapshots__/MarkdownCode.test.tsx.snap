// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MarkdownCode should render code block correctly 1`] = `
<DocumentFragment>
  <pre>
    console.log("hello");
  </pre>
</DocumentFragment>
`;

exports[`MarkdownCode should render inline code correctly 1`] = `
<DocumentFragment>
  <code
    style="color: rgb(255, 255, 255); font-family: monospace; border-radius: 4px;"
  >
    console.log("hello");
  </code>
</DocumentFragment>
`;

exports[`MarkdownCode should render mermaid chart correctly 1`] = `
<DocumentFragment>
  <div>
    <svg />
  </div>
</DocumentFragment>
`;
